api/src/main/java/org/apache/iceberg/expressions/CountNonNull.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.expressions;
+
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.types.Types;
+
+public class CountNonNull<T> extends CountAggregate<T> {
+  private final int fieldId;
+  private final Types.NestedField field;
+
+  protected CountNonNull(BoundTerm<T> term) {
+    super(Operation.COUNT, term);
+    this.field = term.ref().field();
+    this.fieldId = field.fieldId();
+  }
+
+  @Override
+  protected Long countFor(StructLike row) {
+    return term().eval(row) != null ? 1L : 0L;
+  }
+
+  @Override
+  protected Long countFor(DataFile file) {
+    if (field.type().isNestedType()) {
+      return file.recordCount();
```

Comment:
This doesn't seem correct to me. What if it is an optional struct with some number of nulls? I think we may want to return `null` if we don't have a count for the column, which would be the default behavior of `safeGet` right?

Reply:
How about just checking the BoundAggregate in the beginning and not push down if any of the aggregates are on complex type? In this way we don't need to create `AggregateEvaluator` and evaluate any of the aggregates.

Reply:
I'm good ignoring structs, maps, and lists for now. But we _can_ probably handle cases where there are value counts for primitive fields nested within stucts.

To fix this, I don't think that we need to check for struct, list, or map types. Instead, we can simply detect that we don't have the value and null counts (which are not produced for those types) and set the `isValid` flag to signal that we don't know.

Reply:
OK. Thanks